File: routes/status.js
const express = require('express');
const { execFile } = require('child_process');
const util = require('util');
const path = require('path');
const fs = require('fs');
const fsp = fs.promises;

const execFileAsync = util.promisify(execFile);

const router = express.Router();

// GET /api/realtime/ping
router.get('/realtime/ping', async (req, res) => {
  const env_info = {
    PATH: process.env.PATH,
    HOME: process.env.HOME,
    KUBECONFIG: process.env.KUBECONFIG,
    working_directory: process.cwd()
  };

  try {
    const { stdout: whichOut } = await execFileAsync('which', ['kubectl']);
    const kubectlPath = whichOut.trim();

    const { stdout: versionOut } = await execFileAsync('kubectl', ['version', '--client', '-o', 'json']);

    try {
      const version = JSON.parse(versionOut);
      res.json({
        status: 'healthy',
        kubectl: {
          path: kubectlPath,
          version: version.clientVersion?.gitVersion || 'unknown'
        },
        environment: env_info
      });
    } catch (e) {
      res.json({
        status: 'kubectl found but version parse failed',
        kubectl_path: kubectlPath,
        environment: env_info,
        raw_output: versionOut
      });
    }
  } catch (error) {
    res.json({
      status: 'kubectl not available',
      error: error.message,
      environment: env_info,
      stderr: error.stderr || ''
    });
  }
});

// GET /api/snapshot/ping
router.get('/snapshot/ping', async (req, res) => {
  const backupDir = path.join(__dirname, '..', 'k8s-snapshot');
  let available = false;

  try {
    await fsp.access(path.join(backupDir, '.export-complete'));
    available = true;
  } catch {
    // .export-complete doesn't exist — no complete snapshot
  }

  res.json({ available });
});

module.exports = router;

File: cmd/server/routes/status.go
package routes

import (
	"encoding/json"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
)

func registerStatus(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/realtime/ping", handleRealtimePing)
	mux.HandleFunc("GET /api/snapshot/ping", handleSnapshotPing)
	mux.HandleFunc("GET /api/debug/memory", handleDebugMemory)
}

func handleRealtimePing(w http.ResponseWriter, r *http.Request) {
	envInfo := map[string]string{
		"PATH":              os.Getenv("PATH"),
		"HOME":              os.Getenv("HOME"),
		"KUBECONFIG":        os.Getenv("KUBECONFIG"),
		"working_directory": must(os.Getwd()),
	}

	kubectlPath, err := exec.LookPath("kubectl")
	if err != nil {
		writeJSON(w, http.StatusOK, map[string]any{
			"status":      "kubectl not available",
			"error":       err.Error(),
			"environment": envInfo,
		})
		return
	}

	out, err := exec.Command("kubectl", "version", "--client", "-o", "json").Output()
	if err != nil {
		writeJSON(w, http.StatusOK, map[string]any{
			"status":       "kubectl found but version failed",
			"kubectl_path": kubectlPath,
			"environment":  envInfo,
		})
		return
	}

	var versionObj struct {
		ClientVersion struct {
			GitVersion string `json:"gitVersion"`
		} `json:"clientVersion"`
	}
	gitVersion := "unknown"
	if json.Unmarshal(out, &versionObj) == nil && versionObj.ClientVersion.GitVersion != "" {
		gitVersion = versionObj.ClientVersion.GitVersion
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"status": "healthy",
		"kubectl": map[string]string{
			"path":    kubectlPath,
			"version": gitVersion,
		},
		"environment": envInfo,
	})
}

func handleSnapshotPing(w http.ResponseWriter, r *http.Request) {
	// Resolve project root relative to cwd (server runs from project root via go run).
	marker := filepath.Join("k8s-snapshot", ".export-complete")

	_, err := os.Stat(marker)
	writeJSON(w, http.StatusOK, map[string]bool{
		"available": err == nil,
	})
}

func handleDebugMemory(w http.ResponseWriter, r *http.Request) {
	var ms runtime.MemStats
	runtime.ReadMemStats(&ms)
	writeJSON(w, http.StatusOK, map[string]uint64{
		"rss":       ms.Sys / 1024 / 1024,
		"heapUsed":  ms.HeapInuse / 1024 / 1024,
		"heapTotal": ms.HeapSys / 1024 / 1024,
	})
}


// --- helpers ---

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func must(s string, err error) string {
	if err != nil {
		return ""
	}
	return s
}
